### ISSUE TITLE
Data race in sm_enactor.cuh

### ISSUE THREAD

**Issue by AKKamath** [gunrock/gunrock#829]
In sm_enactor.cuh there seems to be a data race between a load and an atomic operation. This occurs on lines 108 and 112.

    -->if (subgraphs[src] >= constrain[0]) {
           return true;
         } else {
           isValid[src] = false;
    -->     atomicAdd(subgraphs + dest, -1);
        }

**Comment by jowens** [comment]
@AKKamath do you have an automated race detector or have you run this code and you found a race and you're pointing it out to us? It's something we should know about either way, but if you've seen a race in practice, that is helpful info for us.

**Comment by AKKamath** [comment]
This data race manifested when I ran the sm application in the example folder with the provided inputs.

**Comment by jowens** [comment]
And you detected it ... with CUDA's tools?

**Comment by Laurawly** [comment]
@AKKamath Thanks for detecting the data race. I should have assigned the atomic results to a different array.

**Comment by jowens** [comment]
@Laurawly Do you think this affects correctness?

**Comment by Laurawly** [comment]
@jowens I don't think this affects correctness. Else condition here is trying to get more related nodes pruned. If the loading `subgraphs[src] >= constrain[0]` happens before the writing `atomicAdd(subgraphs + dest, -1)`, then it works as wished, `src` node could either be kept or pruned in the next iterations. If it's the other case, `src` node could be pruned earlier.

**Comment by jowens** [comment]
OK. Thanks. I understand.

I'd love to see it "fixed" but in the event you don't fix it, could you at least put specific instructions in this issue as to what someone else should do? You know this better than anyone, of course.

**Comment by Laurawly** [comment]
We can use double buffering method. Say in odd iterations, we do loading from buffer1 and writing to buffer2, and in even iterations, we do loading from buffer2 and writing to buffer1.

**Comment by jowens** [comment]
And specifically this is on the array `subgraphs`?

**Comment by Laurawly** [comment]
> And specifically this is on the array `subgraphs`?

Yeah, that's right.
